### PR TITLE
fuzz: fix bug in p2p_headers_presync harness

### PR DESCRIPTION
**Pull request by marcofleon**
The calculation for the test chain's work (`total_work`) should be outside of the loop. Previously, `total_work` was being miscalculated due to multiple additions of work from the same headers. Now, each header's work is only counted once, providing an accurate total.

https://github.com/bitcoin/bitcoin/pull/30918 followup

### CODE REVIEW

**Review by nafeef123 — APPROVED** [review]

**Comment by marcofleon** [comment]
Can be tested by running `FUZZ=p2p_headers_presync ./fuzzbuild/src/test/fuzz/fuzz chainwork_crash.txt`

Should fail on master:
```
Assertion failed: (total_work < chainman.MinimumChainWork()), function p2p_headers_presync_fuzz_target, file p2p_headers_presync.cpp, line 207.
```

and pass on this PR.

[chainwork_crash.txt](https://github.com/user-attachments/files/17151589/chainwork_crash.txt)

**Comment by dergoegge** [comment]
> According to https://oss-fuzz.com/fuzzer-stats?group_by=by-day&fuzzer=afl_bitcoin-core_p2p_headers_presync&project=bitcoin-core the exec/s is 0 for afl.

This is because oss-fuzz doesn't use our afl++ harness but rather the [afl libfuzzer driver](https://github.com/AFLplusplus/AFLplusplus/tree/stable/utils/aflpp_driver). When using this driver, afl++ will fork after the initialization code is run (https://github.com/AFLplusplus/AFLplusplus/blob/d21fb1a558b25c4f46692fa999c0028dfe0eecc0/utils/aflpp_driver/aflpp_driver.c#L393), which might end up not working for harnesses that e.g. create temporary directories or threads during init. You can check the afl++ oss-fuzz logs and you'll see that all inputs in the corpus timeout. This is the case for a few of our harnesses and as consequence they aren't fuzzed with afl++ on oss-fuzz.

File: src/test/fuzz/p2p_headers_presync.cpp [hunk 1]
```diff
@@ -197,15 +197,15 @@ FUZZ_TARGET(p2p_headers_presync, .init = initialize)
...
+    // This is a conservative overestimate, as base is only moved forward when sending headers. In theory,
+    // the longest chain generated by this test is 1600 (FUZZ_MAX_HEADERS_RESULTS * 100) headers. In that case,
+    // this variable will accurately reflect the chain's total work.
+    total_work += CalculateClaimedHeadersWork(all_headers);
 
-        // This test should never create a chain with more work than MinimumChainWork.
-        assert(total_work < chainman.MinimumChainWork());
-    }
+    // This test should never create a chain with more work than MinimumChainWork.
+    assert(total_work < chainman.MinimumChainWork());
```
Comment (dergoegge) [inline src/test/fuzz/p2p_headers_presync.cpp:208]: nit: You could get rid of `total_work` as a variable and just do:

```suggestion
    assert(CalculateClaimedHeadersWork(all_headers) < chainman.MinimumChainWork());
```

Comment (marcofleon) [reply src/test/fuzz/p2p_headers_presync.cpp:208]: If I add the first block to `all_headers` then I could do this. I do like the variable for clarity, and I'd say this assertion is mainly to make the test easier to read. So I'll keep it as is.

**Review by dergoegge — APPROVED** [review]
utACK a7498cc7e26b4b3de976e111de2bd2d79b056b31

**Review by instagibbs — APPROVED** [review]
ACK a7498cc7e26b4b3de976e111de2bd2d79b056b31

**Review by glozow — COMMENTED** [review]
makes sense, utACK a7498cc7e26b4b3de976e111de2bd2d79b056b31

**Review by mzumsande — COMMENTED** [review]
ACK https://github.com/bitcoin/bitcoin/commit/a7498cc7e26b4b3de976e111de2bd2d79b056b31

**Pull request merged**